Copy webServer from Typescript to VS Code

sandersn opened the pull request:
Previously, Typescript packaged the host/session combo needed for running vscode on the web into tsserver.js. tsserver.js would only execute normally if `process` was defined globally. Otherwise it would assume that it was running in a webworker and listen for a startup message. However, after Typescript's module conversion, building tsserver is much simpler if it remains a pure node executable (@jakebailey can explain). The code for the web is a completely separate 500-line chunk that was appended after the bulk of tsserver, so it's easily separable.

This change copies webServer to the typescript language extension and switches vscode to package tsserverlibrary.js for the web instead of tsserver.js. We'll follow it up with a change that removes webServer.ts from tsserver.js in Typescript 5.0, which will slim down our build considerably. It also prepares the way for my update of the web host to use vscode-wasm as a virtual filesystem. It's structured exactly the same way--so it's likely that I'll eventually put the code from my update into the new webServer.ts file from this PR.

Note: Reads better with whitespace ignored.

Some notes:

1. postinstall.mjs now retains tsserverlibrary.js. But tsserver.js is still needed for building for the desktop. @jakebailey will 5.0 remove some of the cost of this duplication? (I'm not actually sure of the cost either)
2. I put webServer.ts in extensions/typescript-language-features/web/webServer.ts and it builds with a separate tsconfig.
3. I switched tsserver.web.js over to use normal webpack. This required touching build infrastructure to support multiple entries in one file.
4. I put all the code into one file without namespaces, which made it easier to manually bundle. I'll probably split it into more than one file at some point, but the file breaks in the Typescript repo didn't make much sense.
5. I refer to lots of internals, which is not great for the long term. We should consider whether some of them should be made public or packaged for public consumption. 
5a. `getLogLevel` is only available on tsserver.js, so I copied its content over. I copied a few trivial things like `const noop = () => {}`.
6. I moved the existing code over mostly unchanged, except to add `ts.server.` prefixes where needed. Vscode's build settings are stricter than typescript's so I had change a couple of places because of that.

sandersn commented:
@mjbvz @jakebailey I think you are the interested parties here.

jakebailey commented:
This is the related TS issue for the package size savings: https://github.com/microsoft/TypeScript/issues/51440

jakebailey commented on extensions/postinstall.mjs:
```diff
@@ -26,7 +26,6 @@ function processRoot() {
 function processLib() {
 	const toDelete = new Set([
 		'tsc.js',
-		'tsserverlibrary.js',
```
I'm curious what this affects; is this then included in the extension? Or is this deleting it from node modules such that it's not there at all?

sandersn replied:
It makes no difference in dist/browser, but (1) I don't have a non-browser build to check (2) it is still in extensions/node_modules/typescript/lib, which might ship with desktop vscode. I don't know.

sandersn replied:
How do I make that happen? 

(And ideally the web would not ship tsserver, just tsserver.web.js, but I expect that's already what happens.)

jakebailey replied:
I'm sorta confused as to why this script modified node_modules at all; that isn't something that would be shipped to end users, right? Is there another step somewhere that is copying the files instead? That seems much clearer to me.

sandersn replied:
> Is there another step somewhere that is copying the files instead?

That is the key question I think.

mjbvz replied:
What we can try is moving this list to a `extensions/.vscodeignore` file. The current approach ensures the dev TS code matches the TS code we'd have in prod, which is nice for confidence in what we are shipping. I guess the other option is to install another full copy of TS as a dev dependency so we can pull in this one file

sandersn replied:
I added to .vscodeignore. It was pretty simple, because the code in postinstall's processLib is way more complicated than the contents of typescript/lib justify; only the 3 files in `toDelete` really matter.

jakebailey commented on extensions/typescript-language-features/extension-browser.webpack.config.js:
```diff
@@ -64,18 +64,20 @@ module.exports = withBrowserDefaults({
 					to: 'typescript/tsserver.web.js',
 					transform: async (content) => {
 						const dynamicImportCompatPath = path.join(__dirname, '..', 'node_modules', 'typescript', 'lib', 'dynamicImportCompat.js');
+						const hostpath = path.join(__dirname, 'out', 'tsServer', 'webServer.js');
```
This seems like an okay hack, but I think long-term this probably just needs to use webServer as the entrypoint and bundle it directly. The webServer file would then do:

```
import * as ts from "typescript/lib/tsserverlibrary"
```

sandersn replied:
I agree, but right now I think it's required because of the dynamic import hack. See the last comment.

mjbvz replied:
With this approach, can we inline the `dynamicImport` logic? I think it is only used in `importPlugin`, right?

sandersn replied:
Yes, it in now inlined as a normal `import` call.

sandersn replied:
tsserver.web.js now uses webServer as its entrypoint.

jakebailey commented on extensions/typescript-language-features/src/tsServer/webServer.ts:
```diff
@@ -0,0 +1,540 @@
+const directorySeparator: string = (ts as any).directorySeparator;
+const combinePaths: (path: string, ...paths: (string | undefined)[]) => string = (ts as any).combinePaths;
+// NOTE: This relies on the current way that vscode injects dynamicImport into ts.server and will probably change
+const dynamicImport: ((id: string) => Promise<any>) | undefined = (ts as any).server.dynamicImport;
```
This doesn't exist on `tsserverlibrary` as it was solely a detail of `tsserver`. You might have to do something special to get the import expression here instead.

sandersn replied:
Assuming that we still ship dynamicImportCompat.js, it's injected at the very start of tsserver.web.js by the webpack hook: 

```js
var ts;
(function (ts) {
    var server;
    (function (server) {
        server.dynamicImport = (id) => import(id);
    })(server = ts.server || (ts.server = {}));
})(ts || (ts = {}));
```

However, I think we have plans to include it tsserverlibrary itself somewhere so this prefix isn't needed. Or maybe work around needing it at all.

jakebailey replied:
dynamicImport is actually defined in the web server code, so would have to be in the code here.

mjbvz commented on extensions/typescript-language-features/src/tsServer/webServer.ts:
```diff
@@ -0,0 +1,539 @@
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+/// <reference types='typescript/lib/tsserverlibrary' />
```
Could the tsserver cod ebe in its own `tsconfig` project? (and own directory under`typescript-language-features`?

mjbvz reviewed (changes requested):
Thank you for looking into this. Added a few comments but excited to see this happening!

sandersn commented on extensions/typescript-language-features/src/utils/platform.ts:
```diff
@@ -5,7 +5,7 @@
 
+declare const navigator: unknown;
 export function isWeb(): boolean {
-	// @ts-expect-error
```
oops. I wondered why this was suddenly an error. I'll put back the comment.

sandersn commented on build/lib/extensions.js:
```diff
@@ -414,19 +414,14 @@ async function webpackExtensions(taskName, isWatch, webpackConfigLocations) {
+                if (outputRoot) {
+                    config.output.path = path.join(outputRoot, path.relative(path.dirname(configPath), config.output.path));
+                }
                 webpackConfigs.push(config);
```
I checked the webpack configs. None of them pass functions, which is good, because this code double-pushed function configs.
Nothing sets `outputRoot` that I could see, so the `if (outputRoot)` code is just-in-case. I left it because I don't know what it's supposed to do.

sandersn commented on extensions/typescript-language-features/extension-browser.webpack.config.js:
```diff
@@ -60,30 +57,20 @@ module.exports = withBrowserDefaults({
+		// packaging depends on that and this must always be like it
+		filename: '[name].js',
+		path: path.join(__dirname, 'dist', 'browser'),
+		libraryTarget: undefined,
```
withBrowserDefaults sets `libraryTarget: commonjs`, which is not correct for tsserver.web.js. I don't know if it's correct for any callers, but there are lots of them so I didn't change it.

jakebailey replied:
It's correct for all callers because they should all be extensions, which even on web are CJS (they are executed in a funky eval environment).

sandersn commented on extensions/typescript-language-features/extension-browser.webpack.config.js:
```diff
@@ -60,30 +57,20 @@ module.exports = withBrowserDefaults({
-					transform: async (content) => {
-						const dynamicImportCompatPath = path.join(__dirname, '..', 'node_modules', 'typescript', 'lib', 'dynamicImportCompat.js');
-						const prefix = fs.existsSync(dynamicImportCompatPath) ? fs.readFileSync(dynamicImportCompatPath) : undefined;
-						const output = await Terser.minify(content.toString());
```
withBrowserDefaults doesn't minify because it sets `mode: none`, but there's a comment that says `mode: production` will minify. So I didn't include a minify entry in tsserver.web.js' config.

sandersn commented:
@mjbvz This is ready for another look. I addressed your comments, and switched away from CopyPlugin in webpack, so can you take a look at the new commits and the follow-up comments I posted with them?

mjbvz commented:
Thank you @sandersn! Let's merge this and test it out further in the next insiders
